### PR TITLE
REF: share DTA/TDA/PA arithmetic methods

### PR DESCRIPTION
**Pull request by jbrockmendel**
This became bigger than I expected, happy to split it into pieces upon request.

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -1082,26 +1087,123 @@ def _cmp_method(self, other, op):
...
+        if other is NaT:
+            # In this case we specifically interpret NaT as a datetime, not
+            # the timedelta interpretation we would get by returning self + NaT
+            result = self.asi8.view("m8[ms]") + NaT.to_datetime64()
+            return DatetimeArray(result)
+
+        i8 = self.asi8
+        result = checked_add_with_arr(i8, other.value, arr_mask=self._isnan)
+        result = self._maybe_mask_results(result)
+        dtype = DatetimeTZDtype(tz=other.tz) if other.tz else DT64NS_DTYPE
```
Comment (mroeschke) [inline pandas/core/arrays/datetimelike.py:1111]: Even with reso support, `DT64NS_DTYPE` is okay since ns is the default?

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:1111]: when the time comes we'll change this to inherit self._reso

Comment (mroeschke) [reply pandas/core/arrays/datetimelike.py:1111]: Worth a TODO or in your local notes?

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:1111]: not really.  I'll be grepping for DT64NS_DTYPE when the time comes anyway

**Review by mroeschke — APPROVED** [review]

**Comment by jreback** [comment]
can you rebase

**Comment by jbrockmendel** [comment]
rebased + green

**Pull request merged**